File: CsharpProjects/myLeetCode/Dynamic_Programming/Subsequence_problem/72. Edit Distance.cs
// 72. Edit Distance
// https://leetcode.com/problems/edit-distance/

#include <iostream>
#include <string>
#include <vector>

int minDistance(std::string word1, std::string word2) {
    int m = word1.length();
    int n = word2.length();

    std::vector<std::vector<int>> dp(m + 1, std::vector<int>(n + 1, 0));

    for (int i = 0; i <= m; i++) {
        dp[i][0] = i;
    }

    for (int j = 0; j <= n; j++) {
        dp[0][j] = j;
    }

    for (int i = 1; i <= m; i++) {
        for (int j = 1; j <= n; j++) {
            if (word1[i - 1] == word2[j - 1]) {
                dp[i][j] = dp[i - 1][j - 1];
            } else {
                dp[i][j] = std::min(dp[i - 1][j - 1], std::min(dp[i - 1][j], dp[i][j - 1])) + 1;
            }
        }
    }

    return dp[m][n];
}

int main() {
    std::vector<std::pair<std::string, std::string>> testCases = {
        {"horse", "ros"},
        {"sunday", "saturday"},
        {"intention", "execution"},
    };

    for (int i = 0; i < testCases.size(); i++) {
        std::string word1 = testCases[i].first;
        std::string word2 = testCases[i].second;

        int distance = minDistance(word1, word2);

        std::cout << "Minimum edit distance for test case " << (i + 1) << ": ";
        std::cout << distance << std::endl;
    }

    return 0;
}

File: C++/代码随想录/Dynamic_Programming/Subsequence_problem/72. Edit Distance.py
def minDistance(word1, word2):
    m, n = len(word1), len(word2)
    dp = [[0] * (n + 1) for _ in range(m + 1)]

    for i in range(m + 1):
        for j in range(n + 1):
            if i == 0:
                dp[i][j] = j  # 如果 word1 为空，需要 j 次插入操作
            elif j == 0:
                dp[i][j] = i  # 如果 word2 为空，需要 i 次删除操作
            elif word1[i - 1] == word2[j - 1]:
                dp[i][j] = dp[i - 1][j - 1]  # 字符相同，不需要操作
            else:
                dp[i][j] = 1 + min(dp[i - 1][j],    # 删除
                                   dp[i][j - 1],    # 插入
                                   dp[i - 1][j - 1]) # 替换

    return dp[m][n]

# 测试用例
print(minDistance("horse", "ros"))  # 输出应该是 3
